tools/PsesLogAnalyzer/Parse-PsesLog.ps1

```diff
@@ -0,0 +1,268 @@
+            $logEntryRegex =
+                [regex]::new(
+                    '^(?<ts>[^\[]+)\[(?<lev>([^\]]+))\]\s+tid:(?<tid>\d+)\s+in\s+''(?<meth>\w+)''\s+(?<file>..[^:]+):\s+line\s+(?<line>\d+)',
+                    [System.Text.RegularExpressions.RegexOptions]::Compiled -bor [System.Text.RegularExpressions.RegexOptions]::IgnoreCase)
+        }
+
+        $filestream =
+            [System.IO.FileStream]::new(
+                $Path,
+                [System.IO.FileMode]:: Open,
+                [System.IO.FileAccess]::Read,
+                [System.IO.FileShare]::ReadWrite,
+                4096,
+                [System.IO.FileOptions]::SequentialScan)
+
+        $streamReader = [System.IO.StreamReader]::new($filestream, [System.Text.Encoding]::UTF8)
+
+        function nextLine() {
+            if ($null -ne $peekBuf) {
+                $line = $peekBuf
+                $script:peekBuf = $null
+            }
+            else {
+                $line = $streamReader.ReadLine()
+            }
+
+            $script:currentLineNum++
+            $line
+        }
+
+        function peekLine() {
+            if ($null -ne $peekBuf) {
+                $line = $peekBuf;
+            }
+            else {
+                $line = $script:peekBuf = $streamReader.ReadLine()
+            }
+
+            $line
+        }
+
+        function parseLogEntryStart([string]$line) {
+            if ($DebugTimingInfo) {
+                $sw = [System.Diagnostics.Stopwatch]::StartNew()
+            }
+
+            while ($line -notmatch $logEntryRegex) {
+                Write-Warning "Ignoring line:${currentLineNum} '$line'"
+                $line = nextLine
+            }
+
+            if (!$HideProgress -and ($script:logEntryIndex % 50 -eq 0)) {
+                Write-Progress "Processing log entry ${script:logEntryIndex} on line: ${script:currentLineNum}"
+            }
+
+            [string]$timestampStr = $matches["ts"]
+            [DateTime]$timestamp = $timestampStr
+            [PsesLogLevel]$logLevel = $matches["lev"]
+            [int]$threadId = $matches["tid"]
+            [string]$method = $matches["meth"]
+            [string]$file = $matches["file"]
+            [int]$lineNumber = $matches["line"]
+
+            $message = parseLogMessage $method
+
+            [PsesLogEntry]::new($script:logEntryIndex, $timestamp, $timestampStr, $logLevel, $threadId, $method,
+                $file, $lineNumber, $message.LogMessageType, $message.LogMessage)
+
+            if ($DebugTimingInfo) {
+                $sw.Stop()
+                if ($sw.ElapsedMilliseconds -gt $DebugTimingThresholdMs) {
+                    Write-Warning "Time to parse log entry ${script:logEntryIndex} - $($sw.ElapsedMilliseconds) ms"
+                }
+            }
+
+            $script:logEntryIndex++
+        }
+
+        function parseLogMessage([string]$Method) {
```

Comment:
Why not a `param` block?

Reply:
Not using this like an advanced function (no need to support pipeline input, should process, etc).